predicate/solver/teleport.py

```diff
@@ -81,6 +81,36 @@ def __and__(self, options: Options):
         """
         return Node(self.expr & options.expr)
 
+@scoped
+class StartSession(ast.Predicate):
+    """
+    StartSession defines the permission to start a moderated session.
+
+    Full documentation here: https://goteleport.com/docs/access-controls/guides/moderated-sessions
+    Note that this predicate models `require_session_join`.
+    """
+
+    mode = ast.String("start_session.mode")
```

Comment:
Should this be a `ast.StringList` instead? We currently support matching multiple modes in the regular system. I'd probably also take a step back here and look at how we could generalize these rules to allow for more complex logic. As it stands this system won't allow for any breakthroughs in customizability here.

Do you think there is a possibility we could model an input as a list of users and use list filter/map/len logic to allow for more customization, not fitting this fairly rigid system?

Reply:
> Should this be a `ast.StringList` instead? We currently support matching multiple modes in the regular system. I'd probably also take a step back here and look at how we could generalize these rules to allow for more complex logic. As it stands this system won't allow for any breakthroughs in customizability here.

`predicate/examples/join_session.py` shows how to specify multiple modes:
```python
((JoinSession.mode == "peer") | (JoinSession.mode == "observer"))
```

Is this what you were thinking?

(In the future we may be able to write the above as `JoinSession.mode in ["peer", "observer"]` (or similar)).

Also, if these modes form a total order (like `Options.recording_mode`), then we would only have to output a single value (the highest one). But from [the docs](https://goteleport.com/docs/access-controls/guides/moderated-sessions/#participant-modes) it seems that there's no TO as, even though `moderator > observer`, `peer` doesn't seem comparable to the other two modes:
- `peer`: Can join and collaborate in a session. They can view output and send input.
- `moderator`: Can join and watch a session. They can view output and forcefully terminate the session at will.
- `observer`: Can join and watch a session. They cannot control the session in any way.


> Do you think there is a possibility we could model an input as a list of users and use list filter/map/len logic to allow for more customization, not fitting this fairly rigid system?

Are you asking about `require_session_join.filter` or in general?